File: kotlin/src/main/kotlin/brackets/brackets.kt
package brackets

import kotlin.collections.ArrayDeque

var openParentheses = mapOf(
  '(' to ')',
  '{' to '}',
  '[' to ']',
)
var closedParentheses = mapOf(
  ')' to '(',
  '}' to '{',
  ']' to '[',
)

@Throws()
fun balancedBrackets(text: String) {
  if (text.isEmpty()) return // short-circuit

  val stack = ArrayDeque<Char>()
  for ((i, c) in text.withIndex()) {
    when {
      c in openParentheses -> {
        stack.push(c)
      }
      c in closedParentheses -> {
        val open = closedParentheses[c]
        val last = stack.removeLastOrNull() ?:
          throw BracketsNotBalancedException("closing bracket $c with no opening bracket at char ${i+1}")
        if (last != open)
          throw BracketsNotBalancedException("closing bracket ] at char ${i+1} mismatched with last opening bracket (")
      }
    }
  }
  if (stack.isNotEmpty()) {
    throw BracketsNotBalancedException("opening brackets without closing brackets found: $stack")
  }
}

class BracketsNotBalancedException(msg: String) : Exception(msg)

fun <E> ArrayDeque<E>.push(item: E) = this.addLast(item)

File: golang/brackets/brackets.go
package brackets

import (
	"fmt"

	"github.com/emirpasic/gods/stacks/linkedliststack"
)

var openParentheses = map[rune]rune{
	'(': ')',
	'{': '}',
	'[': ']',
}
var closedParentheses = map[rune]rune{
	')': '(',
	'}': '{',
	']': '[',
}

func BalancedBrackets(text string) error {
	if text == "" {
		return nil
	}
	runes := []rune(text)
	stack := linkedliststack.New()

	for i, r := range runes {
		if openParentheses[r] != 0 {
			stack.Push(r)
		} else if closedParentheses[r] != 0 {
			open, ok := stack.Pop()
			if !ok {
				return fmt.Errorf("closing bracket %c with no opening bracket at rune %d", r, i+1)
			}
			if open != closedParentheses[r] {
				return fmt.Errorf("closing bracket %c at rune %d mismatched with last opening bracket %c", r, i+1, open)
			}
		}
	}
	if !stack.Empty() {
		stackStrings := "["
		for i, v := range stack.Values() {
			if i <= 0 {
				stackStrings += fmt.Sprintf("%c", v)
			} else {
				stackStrings += fmt.Sprintf(", %c", v)
			}
		}
		stackStrings += "]"
		return fmt.Errorf("opening brackets without closing brackets found: %s", stackStrings)
	}
	return nil
}
func foo() Something {
	return Something{
		Id:     2,
		Name:   "blah",
		Labels: []string{"Q", "R"},
	}
}
